### ISSUE TITLE
[Bug] Weight initialized from the same seed when cpu offloading is active

### ISSUE THREAD

**Issue by francesco-bertolotti** [pytorch/torchtitan#4511]
### Bug description

This does not appear to be a Titan issue per se, but rather stems from a PyTorch bug: https://github.com/pytorch/pytorch/pull/196072

While experimenting with a debug configuration with CPU offloading enabled, I found that training would get stuck during initialization. More specifically, it would hang while initializing the following LM head:

```json
"lm_head": {
    "param_init": {
        "weight": "functools.partial(<function trunc_normal_ at 0x4001540433d0>, std=0.0625, a=-0.1875, b=0.1875)",
        "bias": "<function zeros_ at 0x4001540437f0>"
    }
    ...
}
```

The implementation of `trunc_normal_` that gets executed uses a re-sampling approach: any value falling outside the specified `[a, b]` range is discarded and re-sampled.

However, due to a PyTorch bug, the RNG state is also reset during this process. As a result, `trunc_normal_` cannot make progress toward a valid sample and ends up in an infinite loop.

I’ve opened a [PR](https://github.com/pytorch/pytorch/pull/196072) in PyTorch to address the issue. I’m not entirely sure whether the proposed fix is the best approach, but I wanted to flag the issue and make you aware of it.

----
## Repro config
```python
config = joynet.launcher.Experiment.Config(
    path=pathlib.Path("./data/experiments/llama3-debug"),
    venv=joynet.launcher.Venv.Config(
        path=pathlib.Path(".venv"),
        symlink=True,
    ),
    resources=joynet.launcher.Salloc.Config(
        nnodes=1,
        nproc_per_node=4,
        export={
            "WANDB_MODE": "offline",
        },
        job_name="llama3-debug",
        partition="booster",
        qos="normal",
        time="00:30:00",
        cpus_per_task=288,
        exclusive=True,
    ),
    trainer=torchtitan.trainer.Trainer.Config(
        model_spec=(model := torchtitan.models.llama3.model_registry("debugmodel", attn_backend="varlen")),
        loss=torchtitan.components.loss.ChunkedLossWrapper.Config(
            loss_fn=torchtitan.components.loss.CrossEntropyLoss.Config(
                global_vocab_size=torchtitan.models.common.config_utils.decoder_vocab_size(model),
            ),
        ),
        hf_assets_path="./external/torchtitan/tests/assets/tokenizer",
        optimizer = torchtitan.components.optimizer.OptimizersContainer.Config(
            param_groups=[
                # Exclude the (weight-tied) input/output embedding from weight decay.
                torchtitan.components.optimizer.ParamGroupConfig(
                    pattern=r"(?:tok_embeddings|lm_head)\.",
                    optimizer_name="AdamW",
                    optimizer_kwargs={"lr": 8e-4, "betas": (0.9, 0.95), "eps": 1e-8, "weight_decay": 0.0},
                ),
                # Everything else: standard AdamW with weight decay.
                torchtitan.components.optimizer.ParamGroupConfig(
                    pattern=r".*",
                    optimizer_name="AdamW",
                    optimizer_kwargs={"lr": 8e-4, "betas": (0.9, 0.95), "eps": 1e-8, "weight_decay": 0.1},
                ),
            ],
        ),
        lr_scheduler=torchtitan.components.optimizer.LRSchedulersContainer.Config(
            warmup_steps=2,
            decay_ratio=0.8,
            decay_type="linear",
            min_lr_factor=0.0,
        ),
        training=torchtitan.config.TrainingConfig(
            num_tokens_per_microbatch_per_dp_rank=4 * 2048,
            max_context_length=2048,
            steps=100,
            disable_cuda_graphs=True,
            enable_cpu_offload=True,
        ),
        compile=torchtitan.config.CompileConfig(
            enable=True,
        ),
        dataloader=torchtitan.components.data.GrainDataLoader.Config(
            dataset=torchtitan.components.data.ConcatThenSplitPackingConfig(
                dataset=torchtitan.components.data.SingleDatasetConfig(
                    source=torchtitan.components.data.HuggingFaceRandomAccessSource.Config(
                        path="json",
                        split="train",
                        load_dataset_kwargs={
                            "data_files": "./external/torchtitan/tests/assets/c4_test/data.json",
                        },
                    ),
                    processor=torchtitan.hf_datasets.text_datasets.TextProcessor.Config(),
                    post_filters=(lambda sample: sample is not None,),
                ),
            ),
            shuffle=False,
        ),
        parallelism=torchtitan.config.ParallelismConfig(
            data_parallel_replicate_degree=1,
            data_parallel_shard_degree=-1,
        ),
        checkpoint=torchtitan.components.checkpointer.CheckpointManager.Config(
            enable=True,
            interval=50,
            last_save_model_only=False,
            initial_load_model_only=False,
        ),
        activation_checkpoint=torchtitan.distributed.activation_checkpoint.SelectiveAC.Config(),
        validator=torchtitan.components.validate.Validator.Config(
            freq=5,
            steps=10,
            dataloader=torchtitan.components.data.GrainDataLoader.Config(
                dataset=torchtitan.components.data.ConcatThenSplitPackingConfig(
                    dataset=torchtitan.components.data.SingleDatasetConfig(
                    source=torchtitan.components.data.HuggingFaceRandomAccessSource.Config(
                        path="json",
                        split="train",
                        load_dataset_kwargs={
                            "data_files": "./external/torchtitan/tests/assets/c4_test/data.json",
                        },
                    ),
                    processor=torchtitan.hf_datasets.text_datasets.TextProcessor.Config(),
                    post_filters=(lambda sample: sample is not None,),
                ),
                ),
                shuffle=False,
            ),
        ),
        metrics=torchtitan.components.metrics.MetricsProcessor.Config(
            enable_tensorboard=True,
            enable_wandb=False,
            log_freq=1
        ),
        debug=torchtitan.trainer.DebugConfig(
            save_config_file="config.json",
        ),
    ),
).build().run()
```

## Minimal torch repro
```python
import torch
import torch.distributed as dist
from torch import nn
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.fsdp import CPUOffloadPolicy, fully_shard


@torch.no_grad()
def main():
    assert torch.cuda.is_available(), "this repro requires a CUDA device"
    dist.init_process_group(backend="gloo")

    dp_mesh = init_device_mesh("cuda", (1,))

    lin = nn.Linear(4, 4, bias=False)
    fully_shard(lin, mesh=dp_mesh, offload_policy=CPUOffloadPolicy())

    lin.to_empty(device="cpu")

    for i in range(3):
        lin.weight.normal_()
        print(f"iter {i}: {lin.weight.to_local().flatten()}")

    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

### Versions

```
torch                              2.15.0.dev20260902+cu130
torchvision                        0.30.0.dev20260903+cu130
triton                             3.8.0+gitc01b6774
```

**Comment by tianyu-l** [comment]
cc @albanD
